midp/text_editor.js

```diff
@@ -9,110 +13,138 @@ var TextEditorProvider = (function() {
+            this.setPosition(this.left, this.top);
+            this.setBackgroundColor(this.backgroundColor);
+            this.setForegroundColor(this.foregroundColor);
         },
 
-        setStyle: function(styleKey, styleValue) {
-            // Set input/textarea elem styles.
-            if (!this.styles) {
-                this.styles = {};
-            }
-
-            this.styles[styleKey] = styleValue;
-            if (this.textEditorElem) {
+        _setStyle: function(styleKey, styleValue) {
+            if (this.visible) {
                 this.textEditorElem.style.setProperty(styleKey, styleValue);
             }
         },
 
-        getStyle:  function(styleKey) {
-            return (this.styles && this.styles[styleKey]) || null;
-        },
-
         focus: function() {
             this.focused = true;
-            this.textEditorElem && this.textEditorElem.focus();
+            return new Promise(function(resolve, reject) {
+                if (currentVisibleEditor !== this ||
+                    document.activeElement === this.textEditorElem) {
+                    resolve()
+                    return;
+                }
+                setTimeout(this.textEditorElem.focus.bind(this.textEditorElem));
+                this.textEditorElem.onfocus = resolve;
+            }.bind(this));
         },
 
         blur: function() {
             this.focused = false;
-            this.textEditorElem && this.textEditorElem.blur();
+            return new Promise(function(resolve, reject) {
+                if (currentVisibleEditor !== this ||
+                    document.activeElement !== this.textEditorElem) {
+                    resolve();
+                    return;
+                }
+                setTimeout(this.textEditorElem.blur.bind(this.textEditorElem));
+                this.textEditorElem.onblur = resolve;
+            }.bind(this));
         },
 
         getVisible: function() {
-            return this.visible || false;
+            return this.visible;
         },
 
         setVisible: function(aVisible) {
+            // Check if we need to show or hide the html editor elements.
+            if ((currentVisibleEditor === this && aVisible) ||
+                (currentVisibleEditor !== this && !aVisible)) {
+                this.visible = aVisible;
+                return;
+            }
+
             this.visible = aVisible;
 
             if (aVisible) {
+                if (currentVisibleEditor) {
+                    currentVisibleEditor.visible = false;
+                }
+                currentVisibleEditor = this;
+            } else {
+                currentVisibleEditor = null;
+            }
+
+            if (aVisible) {
+                // Add the "show" class to set opacity to 0.
                 // Sometimes in Java, setVisible() is called after focus(), to make
                 // sure the native input won't lose focus, we change opacity instead
                 // of visibility.
```

Comment:
Hmm, I'm confused, because this code comment suggests that we change opacity instead of visibility to handle the case where *setVisible* is called after *focus*. But in your description of this pull request, you suggest that we need to delay setting focus in that case:

> If the focus is set before the TextEditor is visible, we need to postpone setting focus. Another issue with setting focus is that when calling setVisible and setFocus successively without pause, the html element may not display in time to get focus. So we have to do a setTimeout to let the html element render properly.

Perhaps this code comment is incorrect or obsolete? If so, then I wonder if we should switch back to changing visibility rather than opacity.